odxtools/basicstructure.py

```diff
@@ -31,38 +30,28 @@ class BasicStructure(DopBase):
     parameters: NamedItemList[Parameter]
     byte_size: Optional[int]
 
-    @property
-    def bit_length(self) -> Optional[int]:
+    def get_static_bit_length(self) -> Optional[int]:
         # Explicit size was specified
         if self.byte_size:
             return 8 * self.byte_size
 
-        if all(p.bit_length is not None for p in self.parameters):
-            offset = 0
-            length = 0
-            for param in self.parameters:
-                if isinstance(param, ValueParameter) and hasattr(param.dop, "min_number_of_items"):
-                    # The param repeats itself, making bit_length calculation invalid
-                    # Temporary workaround
-                    # Can not import EndOfPduField to check on its type due to circular dependency
-                    return None
-
-                if param.byte_position is not None:
-                    bit_position_int = param.bit_position if param.bit_position is not None else 0
-                    byte_position_int = (
-                        param.byte_position if param.byte_position is not None else 0)
-                    offset = byte_position_int * 8 + bit_position_int
-                if param.bit_length is None:
-                    return None
-                offset += param.bit_length
-
-                length = max(length, offset)
-
-            # Round up to account for padding bits
-            return math.ceil(length / 8) * 8
-
-        # We were not able to calculate a static bit length
-        return None
+        cursor = 0
+        length = 0
+        for param in self.parameters:
+            param_bit_length = param.get_static_bit_length()
+            if param_bit_length is None:
+                # We were not able to calculate a static bit length
+                return None
+            elif param.byte_position is not None:
+                bit_pos = param.bit_position or 0
+                byte_pos = param.byte_position or 0
+                cursor = byte_pos * 8 + bit_pos
```

Comment:
please double check the spec on this, if I recall correctly, if no byte offset is specified, the max (and not the last) of previous params position+length is used

Reply:
the max is on line 51 (`cursor` is the position of the first bit where the parameter would be encoded)

Reply:
Take this example:
* P1 (len = 4, pos = 2)
* P2 (len = 1, pos = 1)
* P3 (len = 2, pos = None)
The spec says P3 should be encoded from position 6, current code encode it from position 2
Thus total struct size would end up being 4 instead of 8.

Reply:
okay, I looked it up: the spec says "The parameter starts then at the byte edge next to the end of the last parameter."

now it is a matter of interpretation what "last parameter" is supposed to mean. I can see three possibilities:
1) the parameter which has its bytes located furthest back
2) the parameter which we've encountered at that point that has its last coded bit located most to the east
3) the parameter immediately before the currently considered one

the first option does IMO not make any sense at all, the second is what you remember and the third option is the current implementation. I think that the third option is the most likely one, but I would not be too surprised if some vendors went with 2) while others took option 3).

Reply:
In this case, it's fine to choose either, but make sure both encoding logic and bit length calculation logic are using the same logic.

PS: double check the 2.0 spec, it could have different/better wording.

Reply:
> PS: double check the 2.0 spec, it could have different/better wording.

okay, did that. unfortunately this part was copy-and-pasted from 2.0 to 2.2: `Sometimes, the position of a parameter cannot be determined until runtime, [...] the parameter starts then at the byte edge next to the end of the last parameter.`